sim/include/proto/userValue.h

```diff
@@ -0,0 +1,179 @@
+#ifndef BENTOBOX_USERVALUE_H
+#define BENTOBOX_USERVALUE_H
+
+#include <type_traits>
+#include <string>
+#include <bento/protos/values.pb.h>
+
+// Macros for template parameter packs
+#define proto_NUMERIC ::proto::INT32, ::proto::INT64, ::proto::FLOAT32, ::proto::FLOAT64
+
+namespace proto {
+
+// Type aliases for all the protobuf types
+typedef int32_t INT32;
+typedef int64_t INT64;
+static_assert(CHAR_BIT * sizeof(float) == 32);
+typedef float FLOAT32;
+static_assert(CHAR_BIT * sizeof(double) == 64);
+typedef double FLOAT64;
+typedef std::string STR;
+typedef bool BOOL;
+
+// Concept to assert that the given type is a protobuf type
+template <class T>
+concept ProtobufType =
+    std::is_same_v<T, INT32> || std::is_same_v<T, INT64> ||
+    std::is_same_v<T, FLOAT32> || std::is_same_v<T, FLOAT64> ||
+    std::is_same_v<T, STR> || std::is_same_v<T, BOOL>;
+
+// Overloads for setting a value
+void setVal(bento::protos::Value& protoVal, INT32 int32Val);
+void setVal(bento::protos::Value& protoVal, INT64 int64Val);
+void setVal(bento::protos::Value& protoVal, FLOAT32 float32Val);
+void setVal(bento::protos::Value& protoVal, FLOAT64 float64Val);
+void setVal(bento::protos::Value& protoVal, const std::string& strVal);
+void setVal(bento::protos::Value& protoVal, const char* strVal);
+void setVal(bento::protos::Value& protoVal, BOOL boolVal);
+
+// Get value
+template <class T>
+// Ensure that the values are protobuf values
+requires ProtobufType<T> T getVal(const bento::protos::Value& protoVal) {
+    throw std::runtime_error("Invalid protobuf type.");
+}
+
+// Check value type
+template <class T>
+// Ensure that the values are protobuf values
+requires ProtobufType<T> bool isValOfType(
+    const bento::protos::Value& protoVal) {
+    throw std::runtime_error("Invalid protobuf type.");
+}
+
+namespace {
+// Do not export this function. Only isValOfTypes should be used.
+// isValOfTypes does not have the additional N non-type template parameter
+// making invocation simpler.
+template <size_t N = 0, class... AllowedTypes>
+requires(proto::ProtobufType<AllowedTypes>&&...) bool _isValOfTypes(
+    const bento::protos::Value& val) {
+    if constexpr (N == sizeof...(AllowedTypes)) {
+        // Finished looping and no valid types were found
+        return false;
+    } else {
+        // This else statement is required because the compiler seems to
+        // evaluate this typedef first, which will cause an index error
+        // on the iteration where the N == sizeof...(AllowedTypes)
+        typedef std::tuple_element_t<N, std::tuple<AllowedTypes...>>
+            CurrentType;
+        // It matches one of the types
+        if (proto::isValOfType<CurrentType>(val)) {
+            return true;
+        }
+
+        // It doesn't match any type, recurse and get the next type
+        return _isValOfTypes<N + 1, AllowedTypes...>(val);
```

Comment:
Yet again, unfortunately, recursion is necessary. There is no other way to iterate through a list of types. The reason why for loops don't work here is because we're trying to iterate through a list of types (`...AllowedTypes`). When we write a for loop, the iterator value `i` used in a for loop is run-time, not compile time. Hence, we cannot use it to index types:
```c++
for (size_t i = 0; i < sizeof...(AllowedTypes); i++) {
    // This is not allowed since i is not known at compile-time
    typedef std::tuple_element_t<i, std::tuple<AllowedTypes...>> CurrentType;
}```

Reply:
Understood. Can C++17 folding used to replace the recursion instead?

Reply:
I wasn't too familiar with the fold expression, but I got the hang of it now. See 7a55080